Libraries/WebSocket/WebSocket.js

```diff
@@ -11,58 +11,107 @@
  */
 'use strict';
 
-var RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');
-var RCTWebSocketModule = require('NativeModules').WebSocketModule;
+const RCTDeviceEventEmitter = require('RCTDeviceEventEmitter');
+const RCTWebSocketModule = require('NativeModules').WebSocketModule;
+const Platform = require('Platform');
+const WebSocketEvent = require('WebSocketEvent');
 
-var Platform = require('Platform');
-var WebSocketBase = require('WebSocketBase');
-var WebSocketEvent = require('WebSocketEvent');
+const EventTarget = require('event-target-shim');
+const base64 = require('base64-js');
 
-var base64 = require('base64-js');
+import type EventSubscription from 'EventSubscription';
 
-var WebSocketId = 0;
-var CLOSE_NORMAL = 1000;
+const CONNECTING = 0;
+const OPEN = 1;
+const CLOSING = 2;
+const CLOSED = 3;
+
+const CLOSE_NORMAL = 1000;
+
+const WEBSOCKET_EVENTS = [
+  'close',
+  'error',
+  'message',
+  'open',
+];
+
+let nextWebSocketId = 0;
 
 /**
  * Browser-compatible WebSockets implementation.
  *
  * See https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
  * See https://github.com/websockets/ws
  */
-class WebSocket extends WebSocketBase {
+class WebSocket extends EventTarget(WEBSOCKET_EVENTS) {
+  CONNECTING: number = CONNECTING;
+  OPEN: number = OPEN;
+  CLOSING: number = CLOSING;
+  CLOSED: number = CLOSED;
+
+  static CONNECTING = CONNECTING;
+  static OPEN = OPEN;
+  static CLOSING = CLOSING;
+  static CLOSED = CLOSED;
+
   _socketId: number;
-  _subs: any;
+  _subscriptions: Array<EventSubscription>;
+
+  onclose: ?Function;
+  onerror: ?Function;
+  onmessage: ?Function;
+  onopen: ?Function;
+
+  binaryType: ?string;
+  bufferedAmount: number;
+  extension: ?string;
+  protocol: ?string;
+  readyState: number = CONNECTING;
+  url: ?string;
+
+  constructor(url: string, protocols: ?string | ?Array<string>, options: ?{origin?: string}) {
+    super();
+    if (typeof protocols === 'string') {
+      protocols = [protocols];
+    }
 
-  connectToSocketImpl(url: string, protocols: ?Array<string>, headers: ?Object): void {
-    this._socketId = WebSocketId++;
+    if (!Array.isArray(protocols)) {
+      protocols = null;
+    }
 
-    RCTWebSocketModule.connect(url, protocols, headers, this._socketId);
+    this._socketId = nextWebSocketId++;
 
-    this._registerEvents(this._socketId);
+    RCTWebSocketModule.connect(url, protocols, options, this._socketId);
+    this._registerEvents();
   }
 
-  closeConnectionImpl(code?: number, reason?: string): void {
-    this._closeWebSocket(this._socketId, code, reason);
-  }
+  close(): void {
+    if (this.readyState === this.CLOSING ||
+        this.readyState === this.CLOSED) {
+      return;
+    }
 
-  cancelConnectionImpl(): void {
+    this.readyState = this.CLOSING;
     this._closeWebSocket(this._socketId);
```

Comment:
Sure. Can I do that in a follow-up? It'd be a new feature. I wanted to keep this PR here purely mechanical.